### PR TITLE
AP_Scripting: add lua-language-server checks and docs generation.

### PR DESCRIPTION
**Pull request by IamPete1**
This updates the docs and adds configuration and helpers to run lua-language-server. It is smarter than the current luacheck checks we run allowing type checking. https://luals.github.io/ This is the same checker that is used in by the recommended VSCode extension.

The new check is added as part of CI. Lots of stuff currently fails, rather than trying to fix (for the most part) this just adds overrides to ignore warnings in those files. The eventual goal would be to gradually resolve those warnings over time. Most of these failures are relatively harmless, but there are some quite nasty ones in there. 

The tool checks the scripts against the lua documentation file, there are lots of updates to it to ensure that it is correct.

The check can be run locally (after installation) with `python ./Tools/scripts/run_lua_language_check.py` but the recommended way would be to use the VSCode extension which reports the same issues in any case (although we should add the VSCode config to point the extension at the new setup file).

This also enables automatic generation of the docs in a easier to read `.md` format. This can be generated with `./Tools/scripts/generate_lua_docs.sh`. This is added to CI and the docs are uploaded as a artefact.

For example: [ScriptingDocs.md](https://github.com/ArduPilot/ardupilot/files/15228951/ScriptingDocs.md)

The eventual goal would be to convert this to a rst format for direct inclusion into the wiki.

### CODE REVIEW

File: .github/workflows/test_scripting.yml [hunk 1]
```diff
@@ -36,6 +36,27 @@ jobs:
           sudo apt-get -y install lua-check
           ./Tools/scripts/run_luacheck.sh
 
+      - name: Setup language server
+        shell: bash
+        run: |
+          sudo apt install -y curl
+          /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
Comment (tridge) [inline .github/workflows/test_scripting.yml:43]: brew is normally a macos thing .... do we want this?

File: Tools/scripts/generate_lua_docs.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+if [ -n "$(ls -A lualogs)" ]; then
+    echo "lualogs Not Empty"
+    exit 1
+fi
+
+# Need to use abs paths due to lua-language-sever bug
+CONFIG_PATH=$(realpath libraries/AP_Scripting/tests/docs.json)
+DOC_PATH=$(realpath libraries/AP_Scripting/docs/)
+
+lua-language-server --configpath ${CONFIG_PATH}  --logpath lualogs --doc ${DOC_PATH}
```
Comment (tridge) [inline Tools/scripts/generate_lua_docs.sh:16]: does this mean devs need brew installed?

**Review by tridge — CHANGES_REQUESTED** [review]
if this is in CI, and fails PRs if not passing, then alls devs would need to be able to run the checks locally, so will all devs need to install brew ?

**Comment by tridge** [comment]
@IamPete1 maybe we could chat during the week and discuss the developer tooling implications?

**Comment by IamPete1** [comment]
This is the install instructions: https://luals.github.io/#other-install

There are several ways, I used homebrew because I found a example of using it for the install in GitHub actions. I have not tried any of the others. (nor am I really qualified to know which would be best for most people)

**Comment by Ryanf55** [comment]
Nice - looks like this is already part of the recommended VSCode extensions in AP.

**Review by tridge — CHANGES_REQUESTED** [review]
we need a script in Tools/environment_install/install_lua_checker.sh

**Comment by IamPete1** [comment]
```
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:617:17
        Code: need-check-nil
        Need check nil.
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:647:20
        Code: need-check-nil
        Need check nil.
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:627:13
        Code: need-check-nil
        Need check nil.
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:79:28
        Code: param-type-mismatch
        Cannot assign `integer` to parameter `boolean`.
        - `integer` cannot match `boolean`
        - Type `integer` cannot match `boolean`
        - Type `number` cannot match `boolean`
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:261:36
        Code: param-type-mismatch
        Cannot assign `number|nil` to parameter `integer`.
        - `nil` cannot match `integer`
        - Type `nil` cannot match `integer`
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:671:28
        Code: param-type-mismatch
        Cannot assign `nil` to parameter `fun(...any):...unknown`.
        - `nil` cannot match `fun(...any):...unknown`
        - Type `nil` cannot match `fun(...any):...unknown`
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:712:42
        Code: param-type-mismatch
        Cannot assign `string|nil` to parameter `string|number`.
        - `nil` cannot match `string|number`
        - `nil` cannot match any subtypes in `string|number`
        - Type `nil` cannot match `number`
        - Type `nil` cannot match `string`
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:714:39
        Code: param-type-mismatch
        Cannot assign `string|nil` to parameter `string|number`.
        - `nil` cannot match `string|number`
        - `nil` cannot match any subtypes in `string|number`
        - Type `nil` cannot match `number`
        - Type `nil` cannot match `string`
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:617:22
        Code: undefined-field
        Undefined field `read`.
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:647:25
        Code: undefined-field
        Undefined field `read`.
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:627:18
        Code: undefined-field
        Undefined field `seek`.
/home/peter/ardupilot/Tools/AP_Periph/Web/scripts/pppgw_webui.lua:913:22
        Code: redundant-parameter
        This function expects a maximum of 0 argument(s) but instead it is receiving 1.
```

**Review by tridge — APPROVED** [review]

**Pull request merged**